File: tree_walk/optimized/tree_walk.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Find;

my @files;
my $data_dir = -d "tree_walk/_data" ? "tree_walk/_data" : (-d "../../_data" ? "../../_data" : (-d "../_data" ? "../_data" : "_data"));

find(sub {
    push @files, $File::Find::name if -f $_ && /\.txt$/;
}, $data_dir);

my $total_matches = 0;
local $/ = undef;
for my $f (@files) {
    open my $fh, '<:raw', $f or next;
    my $content = <$fh>;
    close $fh;
    my $pos = 0;
    while (($pos = index($content, 'category=', $pos)) != -1) {
        $total_matches++;
        $pos += 9;
    }
}

printf("Tree walk complete: files=%d, matches=%d\n", scalar(@files), $total_matches);

File: tree_walk/optimized/tree_walk.rb
#!/usr/bin/env ruby

def find_files(dir)
  Dir.glob(File.join(dir, "**", "*.txt"))
end

data_dir = "tree_walk/_data"
unless Dir.exist?(data_dir)
  data_dir = if Dir.exist?("../../_data")
               "../../_data"
             elsif Dir.exist?("../_data")
               "../_data"
             elsif Dir.exist?("_data")
               "_data"
             else
               "tree_walk/_data"
             end
end

files = find_files(data_dir)
file_q = SizedQueue.new(500)
res_q = SizedQueue.new(500)
workers = 8

needle = "category=".freeze

threads = Array.new(workers) do
  Thread.new do
    local_matches = 0
    while (batch = file_q.pop)
      break if batch == :done
      batch.each do |path|
        content = File.binread(path)
        pos = 0
        while (pos = content.index(needle, pos))
          local_matches += 1
          pos += 9
        end
      end
    end
    res_q.push(local_matches)
  end
end

prod_t = Thread.new do
  files.each_slice(50) { |slice| file_q.push(slice) }
  workers.times { file_q.push(:done) }
end

total_matches = 0
workers.times do
  total_matches += res_q.pop
end

prod_t.join
threads.each(&:join)

puts "Tree walk complete: files=#{files.size}, matches=#{total_matches}"
